conda/content_trust/__init__.py

```diff
@@ -0,0 +1,238 @@
+            except ValueError:
+                # prefix is not an int and is consequently an invalid file, skip to the next
+                pass
+            else:
+                log.info(f"Loading root metadata from {path}.")
+                trusted = load_metadata_from_file(path)
+                break
+        else:
+            log.debug(
+                f"No root metadata in {context.av_data_dir}. " "Using built-in root metadata."
+            )
+
+        # Refresh trust root metadata
+        while True:
+            # TODO: caching mechanism to reduce number of refresh requests
+            fname = f"{trusted['signed']['version'] + 1}.root.json"
+            path = join(context.av_data_dir, fname)
+
+            try:
+                # TODO: support fetching root data with credentials
+                untrusted = self._fetch_channel_signing_data(
+                    context.signing_metadata_url_base,
+                    fname,
+                )
+
+                verify_root(trusted, untrusted)
+            except HTTPError as err:
+                # HTTP 404 implies no updated root.json is available, which is
+                # not really an "error" and does not need to be logged.
+                if err.response.status_code != 404:
+                    log.error(err)
+                break
+            except Exception as err:
+                # TODO: more error handling
+                log.error(err)
+                break
+            else:
+                # New trust root metadata checks out
+                trusted = untrusted
+                write_metadata_to_file(trusted, path)
+
+        return trusted
+
+    # FUTURE: Python 3.8+, replace with functools.cached_property
+    @property
+    @lru_cache(maxsize=None)
+    def key_mgr(self):
+        trusted = None
+
+        # Refresh key manager metadata
+        fname = KEY_MGR_FILE
+        path = join(context.av_data_dir, fname)
+
+        try:
+            untrusted = self._fetch_channel_signing_data(
+                context.signing_metadata_url_base,
+                KEY_MGR_FILE,
+            )
+
+            verify_delegation("key_mgr", untrusted, self.trusted_root)
+        except (ConnectionError, HTTPError) as err:
+            log.warn(err)
+        except Exception as err:
+            # TODO: more error handling
+            raise
+            log.error(err)
+        else:
+            # New key manager metadata checks out
+            trusted = untrusted
+            write_metadata_to_file(trusted, path)
+
+        # If key_mgr is unavailable from server, fall back to copy on disk
+        if not trusted and exists(path):
+            trusted = load_metadata_from_file(path)
+
+        return trusted
+
+    @staticmethod
+    def _fetch_channel_signing_data(signing_data_url, filename, etag=None, mod_stamp=None):
+        if not context.ssl_verify:
+            warnings.simplefilter("ignore", InsecureRequestWarning)
+
+        session = CondaSession()
```

Comment:
Let's move this into a method, since this may become important in the future when we switch out the fetching backend.

Reply:
You mean move the session definition itself? E.g.

```python
class _SignatureVerification:
    @property
    @lru_cache(maxsize=None)
    def session(self):
        return CondaSession()

    ...
```

Reply:
Yep!